### PR TITLE
Add Java tools deploy jars under third_party/java/java_tools

### PR DESCRIPTION
**Pull request by iirina**
This is part of #7114.

### CODE REVIEW

**Comment by iirina** [comment]
This was pushed.

File: third_party/java/java_tools/update_java_tools.sh [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+For example, to update only JavaBuilder run
+third_party/java/java_tools/update_java_tools.sh JavaBuilder
+
+To update JavaBuilder, Turbine and SingleJar run
+third_party/java/java_tools/update_java_tools.sh JavaBuilder Turbine SingleJar
+"
+
+if [[ ! -z "$1" && $1 = "help" ]]; then
+  echo "$usage"
```
Comment (lberki) [inline third_party/java/java_tools/update_java_tools.sh]: nit: "cat <<EOF" is a bit more idiomatic.

File: third_party/java/java_tools/update_java_tools.sh [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  exit
+fi
+
+# Stores the names of the tools required for update.
+tools_to_update=()
+if [[ ! -z "$@" ]]
```
Comment (lberki) [inline third_party/java/java_tools/update_java_tools.sh]: nit: `if [[ "$#" -eq 0 ]];` is a bit more idiomatic. (IIRC that number doesn't include $0)

File: third_party/java/java_tools/update_java_tools.sh [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+# (which can be either under bazel-bin/ or bazel-genfiles/) under
+# third_party/java/java_tools.
+#
+# Fails if the bazel build fails.
+#
+# bazel_target   The target to be built with bazel.
+# tool_name      The name of the tool associated with the given bazel
+#                target. Used only for printing error messages.
+function update_tool() {
+  local bazel_target="${1}"; shift
```
Comment (lberki) [inline third_party/java/java_tools/update_java_tools.sh]: nit: `"$1"` is enough -- there isn't any need to disambiguate where the name of the variable ends.

File: third_party/java/java_tools/update_java_tools.sh [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  local tool_name="${1}"; shift
+  bazel build "$bazel_target" || (echo "Could not build $tool_name.
+  Please see the Bazel error logs above." && exit 1)
+
+  local binary=$(echo "bazel-bin/$bazel_target" | sed 's@:@/@')
```
Comment (lberki) [inline third_party/java/java_tools/update_java_tools.sh]: nit: wrap `$()` with double quotes. I'd also have written

`"bazel-bin/$(cat "$bazel_target" | sed 's/@:@/@')"`

but that's mostly nitpicking :) (also note that `@` is a valid character in labels, so there are better choices for sed separators)

File: third_party/java/java_tools/update_java_tools.sh [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+  if [[ ! -f "$binary" ]]; then
+    binary=$(echo "bazel-genfiles/$bazel_target" | sed 's@:@/@')
+  fi
+
+  local tool_basename=$(basename $binary)
```
Comment (lberki) [inline third_party/java/java_tools/update_java_tools.sh]: nit: double quotes around `$()` and `$binary` (yes, they nest)

File: third_party/java/java_tools/update_java_tools.sh [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    cp -f "$binary" "third_party/java/java_tools/$tool_basename"
+    updated_tools+=("third_party/java/java_tools/$tool_basename")
+  fi
+}
+
+# Updating the specified tools.
+for tool in "${tools_to_update[@]}"
+do
+  # Get the bazel target associated with the current tool name.
+  tool_bazel_target=${tool_name_to_target[$tool]}
```
Comment (lberki) [inline third_party/java/java_tools/update_java_tools.sh]: nit: double quotes around `${}`

File: third_party/java/java_tools/update_java_tools.sh [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  [[ -z "$tool_bazel_target" ]] && echo "Tool $tool is not supported.
+  Please specify one or more of: ${!tool_name_to_target[*]}." && exit 1
+  update_tool "$tool_bazel_target" "$tool"
+done
+
+if [[ ${#updated_tools[@]} -gt 0 ]]; then
```
Comment (lberki) [inline third_party/java/java_tools/update_java_tools.sh]: Nice! I was considering updating README.md, but overwriting the files in the source control is a somewhat questionable tactic...

**Comment by iirina** [comment]
@lberki this was already pushed 3 days ago (see [my comment](https://github.com/bazelbuild/bazel/pull/7127#issuecomment-455460178)). I'll try to make these changes next time I'm modifying the shell script.

**Comment by lberki** [comment]
I know :) I thought it'd be still useful to look at the code. Better late than never :)

**Pull request closed without merging**